### ISSUE TITLE
2.0.0a5 Multiline `SELECT` are (still) not correctly indented in CTA

### ISSUE THREAD

**Issue by dmohns** [sqlfluff/sqlfluff#4433]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

This is a continuation of https://github.com/sqlfluff/sqlfluff/issues/4234 but I chose to create a new issue instead as this was tested on latest pre-release.

The issue wrong indentation still appears if the queries get more complex. Consider the following three examples (note the slightly different indentation in Example 3)

**Example 1:**
```sql
CREATE TABLE mytable
AS
(SELECT
    id,
    user_id
FROM another_table
)
;
```

**Example 2:**
```sql
CREATE TABLE mytable
AS
(SELECT
    id,
    user_id
FROM another_table
WHERE TRUE
)
;
```

**Example 3:**
```sql
CREATE TABLE mytable
AS
(SELECT
    id,
    user_id
FROM another_table
WHERE 
    TRUE
)
;
```

Examples 1&2 work perfectly, however Example 3 stays as it is.

### Expected Behaviour

See above

### Observed Behaviour

See above

### How to reproduce

See above

### Dialect

BigQuery

### Version

2.0.0a5

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Thanks @dmohns - confirmed on current `main` too (which is basically for the 2.0.0a6 release).

I'll pick this one up, thanks for the excellent example queries!

**Comment by alanmcruickshank** [comment]
I've got a draft fix for this, but it's causing some strange knock-on effects with templated indents. With some luck we'll still get the fix out before 2.0.0.

**Comment by dmohns** [comment]
Hey @alanmcruickshank I'm afraid there is still another variant of this issue around 🙈 

```
CREATE TABLE mytable
AS
(SELECT
    column1,
    column2
FROM table
WHERE TRUE)
;
```

Doesn't raise any lint errors under version `2.0.2`.

Do you want me to re-open this issue or create a new one?

**Comment by alanmcruickshank** [comment]
> Hey @alanmcruickshank I'm afraid there is still another variant of this issue around 🙈 
> 
> ```
> CREATE TABLE mytable
> AS
> (SELECT
>     column1,
>     column2
> FROM table
> WHERE TRUE)
> ;
> ```
> 
> Doesn't raise any lint errors under version `2.0.2`.
> 
> Do you want me to re-open this issue or create a new one?

@dmohns this is currently the intended logic working as expected. The current logic is about matching opening and closing indents. It won't force a new one unless there's an imbalance. In this example - neither opening or closing is taken so there's no imbalance. The equivalent in an expression is that this is also ok:
```sql
select
    ((((((
        8
    ))))))
from foo
```

Do you think we need to update the docs so this is easier to understand?

**Comment by dmohns** [comment]
I will admit, that it took me a while to understand this one 🙈 

What confused me, is that I assumed the only accepted logic is: (roughly speaking)

- Opening bracket is the last character before a line break
- Closing bracket is the first character after a line break

The reason for why I believed that, is because my entire code base get's auto-fixed to that. For example, if we modify the example "slightly" (adding a linebreak basically anywhere in the input) the fix output is 

```sql
CREATE TABLE mytable
AS
(
    SELECT
        column1,
        column2
    FROM table
    WHERE TRUE
);
```

which let me to believe that this is the only accepted way of linting the query.

----

I understand now that logic for linting is actually different, namely, it only cares about the indentation of the _line_ the opening and closing brackets are on, but where in the line the bracket is positioned doesn't matter.

An better example to illustrate this would be

```sql
SELECT * FROM (SELECT
    column1,
    column2
FROM table) WHERE TRUE;
```

**Comment by dmohns** [comment]
Apologies, this is getting pretty long 🙈 There is still some slight ambiguity 

```sql
CREATE TABLE table
AS
(SELECT
    column1,
    column2
FROM table);
```

```sql
SELECT *
FROM
(SELECT
    column1,
    column2
FROM table);
```

The first one is accepted, but the second one isn't. Is there any difference between the `FROM` and `AS`? I guess `enforce_whitespace_preceding_terminator` in the `FromClauseSegment`?
